### DISCUSSION TITLE
require turns into async dependency, why?

### DISCUSSION THREAD

**Discussion by therealgilles** [webpack/webpack#16942] [Q&A]
Hello, I have the code below:
```TSX
  const a = require('m').default();
```
and somehow webpack is turning this into an async dependency. Is there a reason for that? and can I prevent it?
`m` cannot be loaded asynchronous for this to work.

**Comment by alexander-akait** [comment]
Can you show a full code?

**Reply by therealgilles** [reply]
The code looks something like this:
```TSX
const useReactHook = () => {
  if (process.env.VAR) {
    return require('module-a').default();
  } else {
    return require('module-b').default();
  }
};
```

**Reply by therealgilles** [reply]
I've been looking but I see zero documentation on when webpack chooses to use an async dependency instead of the regular webpack require.

**Reply by alexander-akait** [reply]
Why do you mean `async dependency`? Sorry I can't undestand

**Reply by therealgilles** [reply]
I mean `__webpack_async_dependencies__`, similar to the top-level-await example:
```JS
__webpack_require__.a(module, async (__webpack_handle_async_dependencies__) => {
__webpack_require__.r(__webpack_exports__);
/* harmony export */ __webpack_require__.d(__webpack_exports__, {
/* harmony export */   "createUser": () => (/* binding */ createUser)
/* harmony export */ });
/* harmony import */ var _db_connection_js__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./db-connection.js */ 3);
var __webpack_async_dependencies__ = __webpack_handle_async_dependencies__([_db_connection_js__WEBPACK_IMPORTED_MODULE_0__]);
_db_connection_js__WEBPACK_IMPORTED_MODULE_0__ = (__webpack_async_dependencies__.then ? await __webpack_async_dependencies__ : __webpack_async_dependencies__)[0];
```

It is transforming the require into a promise instead of treating it as asynchronous.

**Reply by alexander-akait** [reply]
Can you show me your configuration?

**Reply by therealgilles** [reply]
Unfortunately I cannot. What would turn a require into a dynamic import is what I would like to know.
I'm also seeing webpack doing this with the `uuid` library (v8.3.*):
```JS
... = import('uuid');
```
which is also a dynamic import and I don't know why the `uuid` library would need that. It does not return a promise.

**Reply by alexander-akait** [reply]
Because to load it we need Promises, I don't undestand your problem and why you need this, and so it is hard to help

**Reply by therealgilles** [reply]
Leaving the `uuid` issue aside, `require()` is supposed to be synchronous and a promise is not.
The code example I put above is synchronous, so returning a promise won't work.
`require('module-b').default()` will error out as `default is not a function`.

**Reply by therealgilles** [reply]
@alexander-akait Thank you for trying to help me out. If you have any insight on the different reasons why webpack would transform a require on an asynchronous import, it may help me understand what's going on.

**Reply by alexander-akait** [reply]
Sorry, I really don't undestand you, can you provide example with code where you have a problem, full code, not line, function which I can try to bundle and get the same problem, thank you
